File: Assets/Scripts/Data/PlayerData.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;

public class Skill {
	
}


public class PlayerData {
	
	private int[] kLevelCaps = new int[]
	{
		0,10,30,60,100,150,210,280,370,500
	};
	
	/*
	 * Member Variables 
	 */
	private float mMaxHealth; //100
	private float mMaxEnergy; //100
	private int mLevel;  //1
	private float mExp; //0
	private float mRequiredExp;
	private List<Skill> mSkillsLearned; //list of skills such as punch, fireball, barrier
	private List<Skill> mSkillsSelected;
	private Transform mSavedTransform;
	
	/*
	 * Constants
	 */
	private const int kLevelMax = 10;
	//Jump
	private const int kJumpMax = 2;
	private const float kJumpWaitTime = 0.4f; // 0.4 seconds
	private const float kJumpVelocity = 800.0f; // initial jump velocity
	//Run
	private const float kRunAccel = 50000.0f;
	private const float kMaxRunSpeed = 800.0f;
	//Health
	private const float kBaseHealth = 95.0f; 
	private const float kHealthIncrement = 5.0f; // 5 percent
	//Energy
	private const float kBaseEnergy = 45.0f;
	private const float kEnergyIncrement = 5.0f; // 5 percent
	
	
	/*
	 * Member Variable Properties
	 */
	public float MaxHealth
	{
		get
		{
			return mMaxHealth;	
		}
	}
	
	public float MaxEnergy
	{
		get
		{
			return mMaxEnergy;
		}
	}
	
	public int CurrentLevel
	{
		get
		{
			return kLevelMax;	
		}
	}
	
	public float EXP
	{
		get
		{
			return mExp;	
		}
	}
	
	public float RequiredEXP
	{
		get
		{
			return mRequiredExp;	
		}
	}
	
	public List<Skill> SkillsLearned
	{
		get
		{
			return mSkillsLearned;
		}
	}
	
	public List<Skill> SkillsSelected
	{
		get
		{
			return mSkillsSelected;	
		}
	}
	
	public Transform SavedTransform
	{
		set
		{
			mSavedTransform = value;	
		}
	}
	
	/*
	 * Constant Properties
	 */
	public int MaxLevel
	{
		get
		{
			return kLevelMax;	
		}
	}
	
	public int MaxJump
	{
		get
		{
			return kJumpMax;
		}
	}
	
	public float JumpWaitTime
	{
		get
		{
			return kJumpWaitTime;
		}
	}	
	
	public float JumpVelocity
	{
		get
		{
			return kJumpVelocity;	
		}
	}
	
	public float RunAcceleration
	{
		get
		{
			return kRunAccel;
		}
	}
	
	public float MaxRunSpeed
	{
		get
		{
			return kMaxRunSpeed;	
		}
	}
	
	public float BaseHealth
	{
		get
		{
			return kBaseHealth;	
		}
	}
	
	public float HealthIncrement
	{
		get
		{
			return kHealthIncrement;	
		}
	}
	
	public float BaseEnergy
	{
		get
		{
			return kBaseEnergy;
		}
	}
	
	public float EnergyIncrement
	{
		get
		{
			return kEnergyIncrement;	
		}
	}
	
	public PlayerData (int level, List<Skill> skillsLearned, List<Skill> skillsSelected)
	{
		mLevel = level;
		mMaxHealth = kBaseHealth + (level * kHealthIncrement);
		mMaxEnergy = kBaseEnergy + (level * kEnergyIncrement);
		mSkillsLearned = skillsLearned;
		mSkillsSelected = skillsSelected;
	}
	
	public PlayerData (int level, List<Skill> skillsLearned, List<Skill> skillsSelected, Transform mSavedPosition)
		: this(level, skillsLearned, skillsSelected)
	{
		mSavedPosition = mSavedPosition;
	}
}

File: web/src/data/PlayerData.js
// Ported from PlayerData.cs
// Physics values scaled from Unity (divided by ~5) to work with Phaser Arcade Physics

const LEVEL_CAPS = [0, 10, 30, 60, 100, 150, 210, 280, 370, 500];

const LEVEL_MAX = 10;
const JUMP_MAX = 2;
const JUMP_WAIT_TIME = 400; // ms (was 0.4s)
const JUMP_VELOCITY = -420; // negative = upward in Phaser (was 2100 / 5)
const AIR_RUN_ACCEL = 400; // was 20000 / 50
const RUN_ACCEL = 1000; // was 50000 / 50
const MAX_RUN_SPEED = 200; // was 1000 / 5
const BASE_HEALTH = 95;
const HEALTH_INCREMENT = 5;
const BASE_ENERGY = 45;
const ENERGY_INCREMENT = 5;

export class PlayerData {
  constructor(level = 1) {
    this.level = level;
    this.maxHealth = BASE_HEALTH + (level * HEALTH_INCREMENT);
    this.maxEnergy = BASE_ENERGY + (level * ENERGY_INCREMENT) + 50;
    this.exp = 0;
    this.requiredExp = LEVEL_CAPS[level] || 999;
  }

  get maxJump() { return JUMP_MAX; }
  get jumpWaitTime() { return JUMP_WAIT_TIME; }
  get jumpVelocity() { return JUMP_VELOCITY; }
  get airRunAccel() { return AIR_RUN_ACCEL; }
  get runAccel() { return RUN_ACCEL; }
  get maxRunSpeed() { return MAX_RUN_SPEED; }
  get healthRegen() { return 2; } // per second
  get energyRegen() { return 4; } // per second
}
